File: apps/web/tests/recovery-snapshot.test.mjs
import assert from 'node:assert/strict'
import {readFileSync} from 'node:fs'
import test from 'node:test'
import {createTsLoader} from './helpers/load-ts.mjs'
const core=createTsLoader()('packages/domain-ts/src/recoverySnapshot.ts')
const f=JSON.parse(readFileSync(new URL('../../../contracts/test-vectors/recovery-snapshot-pages.json',import.meta.url)))
const begin=()=>core.beginFullRecoverySnapshot(f.context,'snapshot-a','0')
const page=(items,next=null)=>({snapshotId:'snapshot-a',boundary:'0',items,nextPageToken:next,snapshotComplete:next===null})
test('full snapshot binds pages, completes once and ignores old callbacks',()=>{
  const start=begin(),first=core.stageRecoverySnapshotPage(start,f.context,null,page([f.directory],'next'),1)
  assert.equal(first.stage.complete,false);assert.equal(start.state.access.size,0)
  const last=core.stageRecoverySnapshotPage(first.stage,f.context,'next',page([f.message]),1)
  assert.equal(last.stage.complete,true);assert.equal(last.stage.state.messages.size,1);assert.equal(last.items[0].details.contentType,'image')
  assert.equal(core.stageRecoverySnapshotPage(last.stage,f.context,null,page([f.directory],'next'),1).ignored,true)
  assert.throws(()=>core.stageRecoverySnapshotPage(last.stage,f.context,'unknown',page([])))
})
test('invalid later item discards whole page; changed snapshot and cycling token fail',()=>{
  const first=core.stageRecoverySnapshotPage(begin(),f.context,null,page([f.directory],'next'),1).stage
  assert.throws(()=>core.stageRecoverySnapshotPage(first,f.context,'next',page([f.message,{...f.message,messageId:'b',messageSequence:'3'}])))
  assert.equal(first.state.messages.size,0)
  assert.throws(()=>core.stageRecoverySnapshotPage(first,f.context,'next',page([])))
  assert.throws(()=>core.stageRecoverySnapshotPage(first,f.context,'next',{...page([f.message]),boundary:'1'}))
  assert.throws(()=>core.stageRecoverySnapshotPage(first,f.context,'next',page([f.message],'next'),1))
})
test('revocation removes staged normal bodies from returned items',()=>{
  const result=core.stageRecoverySnapshotPage(begin(),f.context,null,page([f.directory,f.message,f.revoked]))
  assert.equal(result.stage.state.access.get('group:21').readAllowed,false)
  assert.equal(result.items.some(item=>item.kind==='MESSAGE'),false)
})
test('terminal proof cannot regress; same ID cannot move to another sequence',()=>{
  const terminal={...f.message,state:'RECALLED',objectVersion:'2',content:null,details:null}
  const first=core.stageRecoverySnapshotPage(begin(),f.context,null,page([f.directory,terminal],'next'),2).stage
  const last=core.stageRecoverySnapshotPage(first,f.context,'next',page([f.message]))
  assert.equal(last.stage.state.messages.get('message-a').state,'RECALLED');assert.equal(last.items.length,0)
  assert.throws(()=>core.stageRecoverySnapshotPage(first,f.context,'next',page([{...terminal,state:'BURNED'}])))
  assert.throws(()=>core.stageRecoverySnapshotPage(first,f.context,'next',page([{...terminal,messageSequence:'2'}])))
})
test('normal rendering metadata is mandatory, terminal bodies forbidden, empty snapshot valid',()=>{
  for(const details of [{...f.message.details,fromUserId:0},{...f.message.details,createTime:'2026-02-30T00:00:00'},{...f.message.details,isBurn:true}]) {
    assert.throws(()=>core.parseSnapshotItem({...f.message,details}))
  }
  assert.throws(()=>core.parseSnapshotItem({...f.message,state:'BURNED'}))
  assert.equal(core.stageRecoverySnapshotPage(begin(),f.context,null,page([])).stage.complete,true)
})

File: apps/flutter-prototype/test/recovery_snapshot_test.dart
import 'dart:convert';
import 'dart:io';
import 'package:flutter_test/flutter_test.dart';
import 'package:meshx_flutter_probe/core/recovery.dart';
import 'package:meshx_flutter_probe/core/recovery_snapshot.dart';

void main() {
  final f =
      jsonDecode(
            File(
              '../../contracts/test-vectors/recovery-snapshot-pages.json',
            ).readAsStringSync(),
          )
          as Map;
  final c = f['context'];
  final context = RecoveryContext(
    origin: c['origin'],
    userId: c['userId'],
    streamEpoch: c['streamEpoch'],
    generation: c['generation'],
  );
  final directory = Map<String, Object?>.from(f['directory']),
      message = Map<String, Object?>.from(f['message']);
  RecoverySnapshotStage begin() =>
      beginFullRecoverySnapshot(context, 'snapshot-a', '0');
  Map<String, Object?> page(List<Object?> items, [String? next]) => {
    'snapshotId': 'snapshot-a',
    'boundary': '0',
    'items': items,
    'nextPageToken': next,
    'snapshotComplete': next == null,
  };
  final rejects = throwsA(isA<RecoveryProtocolError>());
  test(
    'full snapshot binds pages, completes once and ignores old callbacks',
    () {
      final start = begin(),
          first = stageRecoverySnapshotPage(
            begin(),
            context,
            null,
            page([directory], 'next'),
            limit: 1,
          );
      expect(first.stage.complete, isFalse);
      expect(start.state.access, isEmpty);
      final last = stageRecoverySnapshotPage(
        first.stage,
        context,
        'next',
        page([message]),
        limit: 1,
      );
      expect(last.stage.complete, isTrue);
      expect(last.stage.state.messages.length, 1);
      expect(first.stage.state.messages, isEmpty);
      expect(first.stage.sequences, isEmpty);
      expect(() => last.stage.sequences.clear(), throwsUnsupportedError);
      expect(() => last.stage.positions.clear(), throwsUnsupportedError);
      expect(() => last.stage.consumedTokens.clear(), throwsUnsupportedError);
      expect((last.items.single['details'] as Map)['contentType'], 'image');
      expect(
        stageRecoverySnapshotPage(
          last.stage,
          context,
          null,
          page([directory], 'next'),
          limit: 1,
        ).ignored,
        isTrue,
      );
      expect(
        () =>
            stageRecoverySnapshotPage(last.stage, context, 'unknown', page([])),
        rejects,
      );
    },
  );
  test('invalid later item discards page; wrong boundary and cycle fail', () {
    final first = stageRecoverySnapshotPage(
      begin(),
      context,
      null,
      page([directory], 'next'),
      limit: 1,
    ).stage;
    expect(
      () => stageRecoverySnapshotPage(
        first,
        context,
        'next',
        page([
          message,
          {...message, 'messageId': 'b', 'messageSequence': '3'},
        ]),
      ),
      rejects,
    );
    expect(first.state.messages, isEmpty);
    expect(
      () => stageRecoverySnapshotPage(first, context, 'next', page([])),
      rejects,
    );
    expect(
      () => stageRecoverySnapshotPage(first, context, 'next', {
        ...page([message]),
        'boundary': '1',
      }),
      rejects,
    );
    expect(
      () => stageRecoverySnapshotPage(
        first,
        context,
        'next',
        page([message], 'next'),
        limit: 1,
      ),
      rejects,
    );
  });
  test('revocation removes staged normal bodies from returned items', () {
    final result = stageRecoverySnapshotPage(
      begin(),
      context,
      null,
      page([directory, message, f['revoked']]),
    );
    expect(result.stage.state.access['group:21']!.readAllowed, isFalse);
    expect(result.items.any((item) => item['kind'] == 'MESSAGE'), isFalse);
  });
  test('terminal cannot regress and same ID cannot move sequence', () {
    final terminal = {
      ...message,
      'state': 'RECALLED',
      'objectVersion': '2',
      'content': null,
      'details': null,
    };
    final first = stageRecoverySnapshotPage(
      begin(),
      context,
      null,
      page([directory, terminal], 'next'),
      limit: 2,
    ).stage;
    final last = stageRecoverySnapshotPage(
      first,
      context,
      'next',
      page([message]),
    );
    expect(
      last.stage.state.messages['message-a']!.state,
      MessageRecoveryState.recalled,
    );
    expect(last.items, isEmpty);
    expect(
      () => stageRecoverySnapshotPage(
        first,
        context,
        'next',
        page([
          {...terminal, 'state': 'BURNED'},
        ]),
      ),
      rejects,
    );
    expect(
      () => stageRecoverySnapshotPage(
        first,
        context,
        'next',
        page([
          {...terminal, 'messageSequence': '2'},
        ]),
      ),
      rejects,
    );
  });
  test(
    'normal metadata mandatory, terminal bodies forbidden, empty snapshot valid',
    () {
      final details = Map<String, Object?>.from(message['details'] as Map);
      for (final invalid in [
        {...details, 'fromUserId': 0},
        {...details, 'createTime': '2026-02-30T00:00:00'},
        {...details, 'isBurn': true},
      ]) {
        expect(
          () => parseSnapshotItem({...message, 'details': invalid}),
          rejects,
        );
      }
      expect(() => parseSnapshotItem({...message, 'state': 'BURNED'}), rejects);
      expect(
        stageRecoverySnapshotPage(
          begin(),
          context,
          null,
          page([]),
        ).stage.complete,
        isTrue,
      );
    },
  );
}
